File: packages/prompts/yaml-loader.ts
/*
  Cross-language YAML prompt loader (TypeScript)
  - Requires: yaml
  - Respects PROMPTS_DIR or defaults to packages/prompts
*/
import fs from 'node:fs'
import fsPromises from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import YAML from 'yaml'

export type LoadedPrompt = {
  id: string
  version?: string
  prompt: string
  schema?: unknown
  meta: Record<string, unknown>
}

export type PromptRegistryEntry = {
  id: string
  file: string
  schema?: string
  version?: string
  tags?: string[]
  aliases?: string[]
}

function interpolate( template: string, params: Record<string, unknown> = {} ): string {
  return template.replace( /\{\{\s*([a-zA-Z0-9_]+)\s*\}\}/g, ( _m, key ) => {
    const v = params[key]
    return v === undefined || v === null ? '' : String( v )
  } )
}

function fileExistsSync( p: string ): boolean {
  try {
    fs.accessSync( p )
    return true
  } catch {
    return false
  }
}

async function fileExists( p: string ) {
  try {
    await fsPromises.access( p )
    return true
  } catch {
    return false
  }
}

function resolvePromptsDirSync( custom?: string ): string {
  if ( custom ) return custom
  if ( process.env.PROMPTS_DIR ) return process.env.PROMPTS_DIR

  const moduleDir = path.dirname( fileURLToPath( import.meta.url ) )
  const candidates = [
    path.join( process.cwd(), 'packages', 'prompts' ),
    path.join( process.cwd(), 'prompts' ),
    moduleDir,
  ]

  for ( const p of candidates ) {
    if ( fileExistsSync( path.join( p, 'registry.yaml' ) ) ) return p
  }

  return moduleDir
}

async function resolvePromptsDir( custom?: string ): Promise<string> {
  return resolvePromptsDirSync( custom )
}

function findRegistryEntry(
  entries: PromptRegistryEntry[],
  id: string
): PromptRegistryEntry | undefined {
  return entries.find( ( p ) => p.id === id || ( p.aliases ?? [] ).includes( id ) )
}

function readRegistrySync( dir: string ): PromptRegistryEntry[] {
  const indexPath = path.join( dir, 'registry.yaml' )
  if ( !fileExistsSync( indexPath ) ) {
    throw new Error( `Prompt registry not found: ${indexPath}` )
  }
  const raw = fs.readFileSync( indexPath, 'utf8' )
  const parsed = YAML.parse( raw ) as { prompts: PromptRegistryEntry[] }
  return parsed.prompts
}

async function readRegistry( dir: string ): Promise<PromptRegistryEntry[]> {
  return readRegistrySync( dir )
}

function loadPromptFromEntry(
  dir: string,
  id: string,
  entry: PromptRegistryEntry,
  params?: Record<string, unknown>
): LoadedPrompt {
  const tplPath = path.join( dir, entry.file )
  const tplRaw = fs.readFileSync( tplPath, 'utf8' )
  const tpl = YAML.parse( tplRaw ) as Record<string, unknown>

  const rendered = interpolate( String( tpl.prompt ?? '' ), params )

  let schema: unknown
  const schemaRef = ( tpl.schema_ref as string | undefined ) ?? entry.schema
  if ( schemaRef ) {
    const schemaPath = path.join( dir, schemaRef )
    if ( fileExistsSync( schemaPath ) ) {
      try {
        schema = JSON.parse( fs.readFileSync( schemaPath, 'utf8' ) )
      } catch {
        schema = undefined
      }
    }
  }

  const meta = {
    id: ( tpl.id as string | undefined ) ?? id,
    version: ( tpl.version as string | undefined ) ?? entry.version,
    description: tpl.description,
    owner: tpl.owner,
    tags: tpl.tags,
    runtime: tpl.runtime,
    variables: tpl.variables,
    schema_ref: schemaRef,
    source: path.relative( process.cwd(), tplPath ),
  }

  return { id, version: meta.version as string | undefined, prompt: rendered, schema, meta }
}

export function loadPromptSync(
  id: string,
  params?: Record<string, unknown>,
  options?: { promptsDir?: string }
): LoadedPrompt {
  const dir = resolvePromptsDirSync( options?.promptsDir )
  const idx = readRegistrySync( dir )
  const entry = findRegistryEntry( idx, id )
  if ( !entry ) throw new Error( `Prompt not found in registry: ${id}` )
  return loadPromptFromEntry( dir, id, entry, params )
}

export async function loadPrompt(
  id: string,
  params?: Record<string, unknown>,
  options?: { promptsDir?: string }
): Promise<LoadedPrompt> {
  const dir = await resolvePromptsDir( options?.promptsDir )
  const idx = await readRegistry( dir )
  const entry = findRegistryEntry( idx, id )
  if ( !entry ) throw new Error( `Prompt not found in registry: ${id}` )
  return loadPromptFromEntry( dir, id, entry, params )
}

export async function listPrompts( options?: { promptsDir?: string } ) {
  const dir = await resolvePromptsDir( options?.promptsDir )
  return readRegistry( dir )
}

export function listPromptsSync( options?: { promptsDir?: string } ) {
  const dir = resolvePromptsDirSync( options?.promptsDir )
  return readRegistrySync( dir )
}

File: packages/prompts/yaml_loader.py
"""
Cross-language YAML prompt loader (Python)
- Requires: pip install pyyaml
- Respects PROMPTS_DIR or defaults to repo_root/prompts
"""
from __future__ import annotations
import os
import json
import re
from pathlib import Path
from typing import Any, Dict, Optional

try:
    import yaml  # type: ignore
except Exception as e:  # pragma: no cover
    raise RuntimeError("pyyaml is required: pip install pyyaml") from e


def _interpolate(template: str, params: Optional[Dict[str, Any]] = None) -> str:
    params = params or {}
    pattern = re.compile(r"\{\{\s*([a-zA-Z0-9_]+)\s*\}\}")

    def _repl(match: re.Match[str]) -> str:
        key = match.group(1)
        val = params.get(key)
        return "" if val is None else str(val)

    return pattern.sub(_repl, template)


def _resolve_prompts_dir(custom: Optional[str] = None) -> Path:
    if custom:
        return Path(custom)
    env = os.getenv("PROMPTS_DIR")
    if env:
        return Path(env)

    # Prefer packages/prompts, then prompts at CWD
    cwd = Path.cwd()
    cand_pkg = cwd / "packages" / "prompts"
    if cand_pkg.exists():
        return cand_pkg
    cand_root = cwd / "prompts"
    if cand_root.exists():
        return cand_root

    # Ascend to locate packages/prompts or prompts near repo root
    here = Path(__file__).resolve()
    for base in list(here.parents):
        pkg = base / "packages" / "prompts"
        if pkg.exists():
            return pkg
        plain = base / "prompts"
        if plain.exists():
            return plain

    # Last resort: return expected packages/prompts under CWD (may not exist)
    return cand_pkg


def _find_registry_entry(index: list, prompt_id: str) -> Optional[Dict[str, Any]]:
    for entry in index:
        if entry.get("id") == prompt_id:
            return entry
        aliases = entry.get("aliases") or []
        if prompt_id in aliases:
            return entry
    return None


def list_prompts(prompts_dir: Optional[str] = None) -> Any:
    base = _resolve_prompts_dir(prompts_dir)
    index_path = base / "registry.yaml"
    if not index_path.exists():
        raise FileNotFoundError(f"Prompt registry not found: {index_path}")
    data = yaml.safe_load(index_path.read_text())
    return data.get("prompts", [])


def load_prompt(id: str, params: Optional[Dict[str, Any]] = None, prompts_dir: Optional[str] = None) -> Dict[str, Any]:
    base = _resolve_prompts_dir(prompts_dir)
    index = list_prompts(str(base))
    entry = _find_registry_entry(index, id)
    if not entry:
        raise FileNotFoundError(f"Prompt not found in registry: {id}")

    tpl_path = base / entry.get("file")
    tpl = yaml.safe_load(tpl_path.read_text())

    rendered = _interpolate(str(tpl.get("prompt", "")), params)

    schema = None
    schema_ref = tpl.get("schema_ref") or entry.get("schema")
    if schema_ref:
        schema_path = base / schema_ref
        if schema_path.exists():
            try:
                schema = json.loads(schema_path.read_text())
            except Exception:
                schema = None

    try:
        source = str(tpl_path.relative_to(Path.cwd()))
    except ValueError:
        source = str(tpl_path)

    meta = {
        "id": tpl.get("id", id),
        "version": tpl.get("version", entry.get("version")),
        "description": tpl.get("description"),
        "owner": tpl.get("owner"),
        "tags": tpl.get("tags"),
        "runtime": tpl.get("runtime"),
        "variables": tpl.get("variables"),
        "schema_ref": schema_ref,
        "source": source,
    }

    return {"id": id, "version": meta.get("version"), "prompt": rendered, "schema": schema, "meta": meta}
